src/webui/api/apicontroller.cpp

```diff
@@ -41,8 +41,16 @@ APIController::APIController(ISessionManager *sessionManager, QObject *parent)
 {
 }
 
-QVariant APIController::run(const QString &action, const StringMap &params, const DataMap &data)
+QVariant APIController::run(const QString &action, const StringMap &params, const DataMap &data, ExecutionContext context)
 {
+    const bool isSafe = isActionSafe(action);
+    const bool isUnsafe = isActionUnsafe(action);
```

Comment:
Isn't "unsafe == !safe"?

Reply:
If an action isn't in the "safe" list, we can't determine if it's because it's unsafe or if it's because it doesn't exist. In that case we can't properly throw a 405.

Reply:
>If an action isn't in the "safe" list, we can't determine if it's because it's unsafe or if it's because it doesn't exist. 

1. It's generally bad idea to make things that look mutually exclusive have another meaning.
2. It's generally bad idea to use some "tool" to perform unintended work.
3. It's generally good idea to have default value/behavior/etc. In this case any action can be "unsafe" by default unless it is explicitly declared as "safe" (or vice versa).

>In that case we can't properly throw a 405.

Can't understand why...

Reply:
Sorry, I thought you were staying out of this PR? Unhelpful comments are unhelpful.

Reply:
>Sorry, I thought you were staying out of this PR?

What does it matter?
I just gave you a few basic ideas, the use of which gives a positive result in all cases (except for some very specific ones). If you go against it, you produce illogical and hard to maintain code.
Or am I the only one who sees it as illogical if "unsafe" isn't equal to "not safe"?